components/function-controller/internal/controllers/serverless/build_resources.go

```diff
@@ -172,6 +184,79 @@ func (r *FunctionReconciler) buildService(log logr.Logger, instance *serverlessv
 			},
 		},
 	}
+}
+
+func (r *FunctionReconciler) buildInternalImageAddress(instance *serverlessv1alpha1.Function) string {
+	imageTag := r.calculateImageTag(instance)
+	return fmt.Sprintf("%s/%s-%s:%s", r.config.Docker.Address, instance.Namespace, instance.Name, imageTag)
+}
+
+func (r *FunctionReconciler) buildExternalImageAddress(instance *serverlessv1alpha1.Function) string {
+	imageTag := r.calculateImageTag(instance)
+	return fmt.Sprintf("%s/%s-%s:%s", r.config.Docker.ExternalAddress, instance.Namespace, instance.Name, imageTag)
+}
+
+func (r *FunctionReconciler) sanitizeDependencies(dependencies string) string {
+	result := "{}"
+	if strings.Trim(dependencies, " ") != "" {
+		result = dependencies
+	}
 
-	return service
+	return result
+}
+
+func (r *FunctionReconciler) functionLabels(instance *serverlessv1alpha1.Function) map[string]string {
+	labels := make(map[string]string, len(instance.GetLabels())+3)
+	for key, value := range instance.GetLabels() {
+		labels[key] = value
+	}
+
+	return r.mergeLabels(labels, r.internalFunctionLabels(instance))
+}
```

Comment:
It doesn't make sense here to copy labels from instance to separate map, smiply:
```suggestion
func (r *FunctionReconciler) functionLabels(instance *serverlessv1alpha1.Function) map[string]string {
	return r.mergeLabels(instance.GetLabels(), r.internalFunctionLabels(instance))
}
```